velox/functions/prestosql/window/FirstLastValue.cpp

```diff
@@ -0,0 +1,145 @@
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool, nullptr) {
+    VELOX_CHECK_NULL(args[0].constantValue);
+    columnIndex_ = args[0].index.value();
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      const SelectivityVector& validRows,
+      int32_t resultOffset,
+      const VectorPtr& result) override {
+    auto numRows = frameStarts->size() / sizeof(vector_size_t);
+    std::vector<vector_size_t> rowNumbers(numRows, -1);
+
+    if constexpr (TValue == ValueType::kFirst) {
+      auto frameStartsPtr = frameStarts->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers[i] = frameStartsPtr[i]; });
+    } else {
+      auto frameEndsPtr = frameEnds->as<vector_size_t>();
+      validRows.applyToSelected(
+          [&](auto i) { rowNumbers[i] = frameEndsPtr[i]; });
+    }
+
+    auto rowNumbersRange = folly::Range(rowNumbers.data(), numRows);
+    partition_->extractColumn(
+        columnIndex_, rowNumbersRange, resultOffset, result);
+  }
+
+ private:
+  // Index of the first_value / last_value argument column in the input row
+  // vector. This is used to retrieve column values from the partition data.
+  column_index_t columnIndex_;
+
+  const exec::WindowPartition* partition_;
+};
+
+template <TypeKind kind>
+std::unique_ptr<exec::WindowFunction> createFirstLastValueFunction(
+    const std::vector<exec::WindowFunctionArg>& args,
+    const TypePtr& resultType,
+    velox::memory::MemoryPool* pool,
+    const ValueType valueType) {
+  using T = typename TypeTraits<kind>::NativeType;
+  if (valueType == ValueType::kFirst) {
+    return std::make_unique<FirstLastValueFunction<ValueType::kFirst, T>>(
+        args, resultType, pool);
+  }
+  return std::make_unique<FirstLastValueFunction<ValueType::kLast, T>>(
+      args, resultType, pool);
+}
+
+} // namespace
+
+template <ValueType TValue>
+void registerFirstLastInternal(const std::string& name) {
+  // first_value(T) -> T
+  // last_value(T) -> T
+  std::vector<exec::FunctionSignaturePtr> signatures{
+      exec::FunctionSignatureBuilder()
+          .typeVariable("T")
+          .returnType("T")
+          .argumentType("T")
+          .build(),
+  };
+
+  exec::registerWindowFunction(
+      name,
+      std::move(signatures),
+      [name](
+          const std::vector<exec::WindowFunctionArg>& args,
+          const TypePtr& resultType,
+          velox::memory::MemoryPool* pool,
+          HashStringAllocator* /*stringAllocator*/)
+          -> std::unique_ptr<exec::WindowFunction> {
+        auto typeKind = args[0].type->kind();
+        if constexpr (TValue == ValueType::kFirst) {
```

Comment:
Just curious : Is this if condition required ? Can we directly pass the TValue in the DISPATCH as a parameter.